src/libstd/sys/unix/stack_overflow.rs

```diff
@@ -135,26 +135,44 @@ mod imp {
         Handler { _data: MAIN_ALTSTACK };
     }
 
-    pub unsafe fn make_handler() -> Handler {
-        let alt_stack = mmap(ptr::null_mut(),
-                             SIGSTKSZ,
-                             PROT_READ | PROT_WRITE,
-                             MAP_PRIVATE | MAP_ANON,
-                             -1,
-                             0);
-        if alt_stack == MAP_FAILED {
+    #[cfg(any(target_os = "linux",
+              target_os = "macos",
+              target_os = "bitrig",
+              target_os = "netbsd",
+              target_os = "openbsd"))]
+    unsafe fn get_stack() -> libc::stack_t {
+        let stackp = mmap(ptr::null_mut(),
+                          SIGSTKSZ,
+                          PROT_READ | PROT_WRITE,
+                          MAP_PRIVATE | MAP_ANON,
+                          -1,
+                          0);
+        if stackp == MAP_FAILED {
             panic!("failed to allocate an alternative stack");
         }
+        libc::stack_t { ss_sp: stackp, ss_flags: 0, ss_size: SIGSTKSZ }
+    }
 
-        let mut stack: libc::stack_t = mem::zeroed();
+    #[cfg(any(target_os = "dragonfly",
+              target_os = "freebsd"))]
+    unsafe fn get_stack() -> libc::stack_t {
+        let stackp = mmap(ptr::null_mut(),
+                          SIGSTKSZ,
+                          PROT_READ | PROT_WRITE,
+                          MAP_PRIVATE | MAP_ANON,
+                          -1,
+                          0);
+        if stackp == MAP_FAILED {
+            panic!("failed to allocate an alternative stack");
+        }
+        libc::stack_t { ss_sp: stackp as *mut i8, ss_flags: 0, ss_size: SIGSTKSZ }
```

Comment:
The reason this is duplicated is because there is only one mmap and it always returns a *mut libc::c_void.  On FreeBSD and DragonflyBSD, the libc::stack_t.ss_sp member is a *mut i8.  Also, because we have compile flags that say treat all warnings as errors, the FreeBSD and DragonflyBSD builds were failing.

What is needed is a conditional compilation block that casts the result of mmap to a *mut i8 when building for FreeBSD and DragonflyBSD.  BUT, we don't have conditional compilation for general expressions yet.  We only have it for functions.  So I'm forced to wrap the mmap call and libc::stack_t initialization in a function that can be conditionally compiled based on OS.